Added support for CompletionStage and extensions

FroMage opened the pull request:
This is a preview PR for #46

Don't merge it until https://github.com/wildfly/jandex/pull/43 is merged and jandex is released. However, this should support:

- the JAX-RS 2.1 return type of `CompletionStage`
- the ability to register extensions to the scanning process for:
    - custom types such as RESTEasy `@*Param` that use the method parameter name
    - custom async return types such as RESTEasy support for RxJava 1&2 `Single`

Please review so we can get this merged ASAP when Jandex is released. Thanks.

EricWittmann commented on implementation/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiAnnotationScanner.java:
```diff
@@ -128,6 +131,8 @@
  */
 public class OpenApiAnnotationScanner {
 
+    private static final DotName COMPLETION_STAGE_NAME = DotName.createSimple(CompletionStage.class.getName());
```
Maybe put this in https://github.com/smallrye/smallrye-open-api/blob/master/implementation/src/main/java/io/smallrye/openapi/api/OpenApiConstants.java

FroMage replied:
Done

EricWittmann commented on implementation/src/main/java/io/smallrye/openapi/runtime/scanner/AnnotationScannerExtension.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Extension point for supporting extensions to JAX-RS.
+ */
+public class AnnotationScannerExtension {
```
This is a class rather than an interface?

FroMage replied:
Yes, so that extensions don't have to implement every method.

EricWittmann replied:
I wonder if an interface with default methods might better capture the intent.

FroMage replied:
Well, yes that would work too. I guess it only depends on your minimum Java version, but if it's fine by you I can change that.

msavy replied:
+1 to interface. AFAIK, (cc @kenfinnigan), Smallrye expects Java >=8.

kenfinnigan replied:
Correct, right now it's Java 8+

EricWittmann replied:
Great, then I think the consensus is interface with default methods.  Otherwise I think this looks good and we just need to wait for jandex before merging.  Thanks for the contribution!

EricWittmann commented:
This looks interesting.  Am I right in assuming that it's up to the platform to load the extensions in whatever way makes sense?  So e.g. Thorntail would be responsible for loading the extensions in some Thorntaily way?  And other platforms would do the same.

FroMage commented:
ATM my usage doesn't need this to be a service. It could make sense to also allow for this to be service-loaded. In general I try to not add features before there's a user for them.

EricWittmann commented:
This change looks great to me - with potential for additional extension points in the future.  So a big +1 from me on this.  Just let me know when it's ready to be merged (when the new version of jandex is available).

FroMage commented:
I rebased, added an extension point, depend on latest released jandex and turned the class into an interface, but that makes the implementation really ugly because you can't use `super` to invoke the super class to avoid figuring out what is the right default return value, you need to call `return return AnnotationScannerExtension.super.resolveAsyncType(type);` which IMO is much worse than if we were using classes as initially.

FroMage commented:
It'd be really great if we could get this reviewed and released ASAP because we need it in another project :)

msavy commented on implementation/src/main/java/io/smallrye/openapi/runtime/scanner/AnnotationScannerExtension.java:
```diff
@@ -0,0 +1,66 @@
+     * @return the parameter location, or null if unknown
+     */
+    public default In parameterIn(MethodParameterInfo paramInfo) {
+        return null;
```
I might be missing something, but why are you providing default/null implementations in this interface instead of just a non-default definition and requiring the user to provide a null impl if they really want that.

This might relate to your comment https://github.com/smallrye/smallrye-open-api/pull/50#issuecomment-450897205 in a way I'm not grokking.

FroMage replied:
Because providing a default impl makes it instantly obvious to users what a valid default return value is. This shows that returning `null` is acceptable and will allow other extensions to participate. Otherwise you have to rely on javadoc and have no code to prove it.

msavy replied:
Okay, thanks, I wasn't sure whether an implementor should generally be expected to provide an implementation to all these methods for it to be sensible.

FroMage commented:
@kenfinnigan it's up to you guys. If it weren't for that `super` annoyance, I would not have minded either way, but that difference probably means most extension implementors will never delegate to the superinterface because they won't know how to invoke that super method. I knew it was possible only because I followed the implementation of default methods in Java at the time they were introduced, which is a pretty high bar…

EricWittmann commented:
If the usage is ugly perhaps an interface without default implementations and instead include a default/abstract/no-op implementation class?  I honestly don't have a strong opinion on this, so I'm fine with whatever the consensus might be. :)

kenfinnigan commented:
@FroMage if you want to simplify the implementation usage, then I'm good with whatever change that means

FroMage commented:
OK I just pushed a change to have an interface and default base class implementing that. I still don't see why that's better than a single class, but I don't care at all, for implementors it's still extending a single class and no `super` weirdness.

FroMage commented:
Thanks! When do you think we can expect a release? Even a Beta would be good :)

kenfinnigan commented:
In terms of a release, we may have a slight problem.

Current master has been updated to work against 1.1-RC1 of the spec, and the auto release process we currently have doesn't support releasing anything containing a "Beta" type designation. I think you have admin access to Semaphore so you can take a look and see what I mean

EricWittmann commented:
@FroMage What sort of timeframe do you need for a release?  We should be releasing MP OpenAPI 1.1 in the next week or so.  Do you need this released sooner than, say, two weeks from now?

FroMage commented:
Two weeks should be good, but if you can't release with Jandex Beta that's another problem to solve…

kenfinnigan commented:
It would be better if it wasn't Beta, but not sure what Jason's plans around that are

EricWittmann commented:
@FroMage and @kenfinnigan - Version 1.1.1 of MP+OAI has been released and I have updated SmallRye OpenAPI to support it.  Everything looks good from my perspective.  I'd be happy if a release were done asap.  Any chance jandex has a Final release out that can be used?

FroMage commented:
I think @n1hility and @stuartwdouglas can make jandex releases.

FroMage commented:
Yay, thanks!

kenfinnigan commented:
Sounds good @EricWittmann. Do you need me to pull trigger or you can?

EricWittmann commented:
I don't know how.  But if you point me in the right direction I'm happy to do it. :)

kenfinnigan commented:
Pulled the trigger, so will build shortly

kenfinnigan commented:
@FroMage, @EricWittmann - https://search.maven.org/search?q=smallrye-open-api

1.1.0 is now in Central

FroMage commented:
Great, thanks!
